contracts/multisig/src/types.rs

```diff
@@ -2,31 +2,85 @@ use std::{collections::HashMap, fmt};
 
 use axelar_wasm_std::Snapshot;
 use cosmwasm_schema::cw_serde;
-use cosmwasm_std::{from_binary, Addr, HexBinary, StdResult};
+use cosmwasm_std::{from_binary, Addr, HexBinary, StdError, StdResult};
 use cw_storage_plus::{KeyDeserialize, PrimaryKey};
 
-use crate::ContractError;
+use crate::{secp256k1::ecdsa_verify, ContractError};
 
 pub trait VerifiableSignature {
     fn verify(&self, msg: &MsgToSign, pub_key: &PublicKey) -> Result<bool, ContractError>;
 }
 
 #[cw_serde]
-pub struct PublicKey(HexBinary);
+pub enum KeyType {
```

Comment:
Discussed with Chris as well, here's what we came up with to make these nicer. Warning: pseudocode.
```
enum KeyType {
  ECDSA,
  ED25519,
}

enum Signature {
  ECDSA(HexBinary)
  ED25519(HexBinary)
}

enum PublicKey {
    ECDSA(HexBinary)
    ED25519(HexBinary)
}

trait KeyTyped {
  fn matches<T>(&self, other: T) -> bool where T: KeyTyped {
    self.key_type() == other.key_type()
  };

  fn key_type(&self) -> KeyType;
}

impl KeyTyped for PublicKey {
  fn key_type(&self) -> KeyType {
    match self {
      ECDSA(_) => KeyType::ECDSA,
      ED25519(_) => KeyType::ED25519,
    }
  }
}

impl KeyTyped for Signature {
  fn key_type(&self) -> KeyType {
    match self {
      ECDSA(_) => KeyType::ECDSA,
      ED25519(_) => KeyType::ED25519,
    }
  }
}

impl Signature {
  pub fn verify(&self, pub_key: PublicKey) -> Result<bool, Error> {
    if !self.matches(pub_key) {
      return Err(Error::new("Key type mismatch"));
    }

    match self {
      ECDSA(sig) => {
        todo!()
      },
      ED25519(sig) => {
        todo!()
      },
    }
  }
}
```

Reply:
@cjcobb23 Also, you need to check out this https://serde.rs/enum-representations.html. I personally prefer it being internally tagged, i.e. `#[serde(tag = "type")]`.

Reply:
I implemented this, take a look